File: scripts/quality/run-auto-pdq.py
#!/usr/bin/env python3
"""Run PIDEX quality cadence automatically after terminal pipeline events.

Fail-soft helper for scripts/pipeline/event.sh. It writes a PDQ report for the
completed plan, updates review-state, and appends an OpQualityReview event.
"""
from __future__ import annotations

import argparse
import datetime as dt
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
STATE = ROOT / "state"


def slug(value: str) -> str:
    return (re.sub(r"[^A-Za-z0-9._-]+", "-", value).strip("-") or "unknown")[:100]


def normalize_plan(value: str) -> str:
    raw = (value or "unknown-plan").strip() or "unknown-plan"
    m = re.fullmatch(r"(?:plan-)?(\d{1,3})", raw, re.I)
    if m:
        return f"plan-{m.group(1).zfill(3)}"
    m = re.match(r"(?:plan-)?(\d{1,3})[-_]", raw, re.I)
    if m:
        return f"plan-{m.group(1).zfill(3)}"
    return raw


def append_quality_operator_event(project: Path, plan_key: str, pipeline_id: str, report_payload: dict[str, Any], terminal_event: str) -> Path:
    project_path = str(project.expanduser().resolve())
    out_dir = STATE / "orchestrator-events" / slug(project_path)
    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / f"{slug(pipeline_id)}.jsonl"
    rec = {
        "timestamp": dt.datetime.now(dt.timezone.utc).isoformat(),
        "project_path": project_path,
        "project_slug": project.name or slug(project_path),
        "pipeline_id": slug(pipeline_id),
        "plan_key": plan_key,
        "operator_type": "OpQualityReview",
        "actor": "orchestrator",
        "source": "auto-pdq",
        "logical_decision": {
            "trigger": terminal_event,
            "review_scope": "plan",
            "update_review_state": True,
        },
        "physical_action": {
            "json_report": report_payload.get("json"),
            "markdown_report": report_payload.get("markdown"),
            "review_state": report_payload.get("review_state"),
        },
        "confidence": report_payload.get("confidence"),
        "trace_gaps": report_payload.get("trace_gaps"),
        "plans_reviewed": report_payload.get("plans", []),
    }
    with out_path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(rec, separators=(",", ":"), sort_keys=True) + "\n")
    return out_path


def main() -> int:
    ap = argparse.ArgumentParser(description="Run automatic PDQ for a terminal pipeline event")
    ap.add_argument("--project", required=True)
    ap.add_argument("--plan", required=True)
    ap.add_argument("--pipeline-id", required=True)
    ap.add_argument("--terminal-event", required=True)
    args = ap.parse_args()

    project = Path(args.project).expanduser().resolve()
    plan_key = normalize_plan(args.plan)
    script = ROOT / "scripts" / "quality" / "report.py"
    proc = subprocess.run(
        [sys.executable, str(script), "--project", str(project), "--plan", plan_key, "--update-review-state"],
        cwd=str(ROOT),
        text=True,
        capture_output=True,
        timeout=120,
    )
    if proc.returncode != 0:
        sys.stderr.write(proc.stderr or proc.stdout)
        return proc.returncode or 1
    try:
        payload = json.loads(proc.stdout)
    except Exception as exc:
        sys.stderr.write(f"failed to parse PDQ output: {exc}\n{proc.stdout}\n")
        return 1
    event_path = append_quality_operator_event(project, plan_key, args.pipeline_id, payload, args.terminal_event)
    print(json.dumps({
        "auto_pdq": True,
        "json": payload.get("json"),
        "markdown": payload.get("markdown"),
        "review_state": payload.get("review_state"),
        "operator_event": str(event_path),
        "confidence": payload.get("confidence"),
        "trace_gaps": payload.get("trace_gaps"),
    }, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/quality/run-auto-pdq.mjs
#!/usr/bin/env node
import { existsSync, mkdirSync, writeFileSync } from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { spawn, spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';
const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..');
const STATE = path.join(ROOT, 'state');
function parse(argv) { const a = { project: ROOT, plan: 'unknown-plan', pipelineId: '', terminalEvent: 'pipeline_completed' }; for (let i = 0; i < argv.length; i++) { if (argv[i].startsWith('--')) a[argv[i].slice(2).replace(/-([a-z])/g, (_, c) => c.toUpperCase())] = argv[++i]; } return a; }
function slug(value) { return String(value || 'unknown').replace(/[^A-Za-z0-9._-]+/g, '-').replace(/^-+|-+$/g, '').slice(0, 100) || 'unknown'; }
function normalizePlan(value) { const raw = String(value || 'unknown-plan').trim() || 'unknown-plan'; let m = raw.match(/^(?:plan-)?(\d{1,3})$/i); if (m) return `plan-${m[1].padStart(3, '0')}`; m = raw.match(/^(?:plan-)?(\d{1,3})[-_]/i); if (m) return `plan-${m[1].padStart(3, '0')}`; return raw; }
function runContractGovernor(project, plan, reportOut) {
  try {
    if (['0', 'false', 'no', 'off'].includes(String(process.env.PIDEX_CONTRACT_GOVERNOR || '0').toLowerCase())) return;
    if (process.env.PIDEX_CONTRACT_GOVERNOR_SOURCE === 'governor-refresh') return;
    const report = reportOut?.json;
    if (!report) return;
    const script = path.join(ROOT, 'scripts', 'quality', 'contract-governor.mjs');
    if (!existsSync(script)) return;
    const child = spawn(process.execPath, [script, 'run', '--project', path.resolve(project), '--plan', normalizePlan(plan), '--report', report], {
      cwd: ROOT,
      detached: true,
      stdio: 'ignore',
      env: { ...process.env, PIDEX_CONTRACT_GOVERNOR_SOURCE: 'auto-pdq' },
    });
    child.unref();
  } catch {}
}
function appendQualityReview(project, plan, pipelineId, terminalEvent, reportOut) {
  const projectPath = path.resolve(project);
  const planKey = normalizePlan(plan);
  const pid = slug(pipelineId || `${slug(projectPath)}-${slug(planKey)}`);
  const rec = {
    timestamp: new Date().toISOString(),
    project_path: projectPath,
    project_slug: path.basename(projectPath) || slug(projectPath),
    pipeline_id: pid,
    plan_key: planKey,
    actor: 'orchestrator',
    source: 'auto-pdq',
    operator_type: 'OpQualityReview',
    plans_reviewed: Array.isArray(reportOut?.plans) ? reportOut.plans : [planKey],
    confidence: reportOut?.confidence || 'descriptive-only',
    trace_gaps: Number.isFinite(Number(reportOut?.trace_gaps)) ? Number(reportOut.trace_gaps) : null,
    logical_decision: { review_scope: 'plan', trigger: terminalEvent || 'pipeline_completed', update_review_state: Boolean(reportOut?.review_state) },
    physical_action: { json_report: reportOut?.json || null, markdown_report: reportOut?.markdown || null, review_state: reportOut?.review_state || null }
  };
  const dir = path.join(STATE, 'orchestrator-events', slug(projectPath));
  mkdirSync(dir, { recursive: true });
  const file = path.join(dir, `${pid}.jsonl`);
  writeFileSync(file, `${JSON.stringify(rec)}\n`, { encoding: 'utf8', flag: 'a' });
  return file;
}
const a = parse(process.argv.slice(2));
if (['0', 'false', 'no', 'off'].includes(String(process.env.PIDEX_AUTO_PDQ || '1').toLowerCase())) process.exit(0);
const script = path.join(ROOT, 'scripts', 'quality', 'report.mjs');
if (!existsSync(script)) process.exit(0);
const cp = spawnSync(process.execPath, [script, '--project', a.project, '--plan', a.plan], { cwd: ROOT, encoding: 'utf8', timeout: Number(process.env.PIDEX_AUTO_PDQ_TIMEOUT_SECONDS || 120) * 1000 });
const stdout = cp.stdout.trim();
if (stdout) console.log(stdout);
if (cp.status !== 0) { if (cp.stderr.trim()) console.error(cp.stderr.trim()); process.exit(cp.status || 1); }
try {
  const parsed = stdout ? JSON.parse(stdout) : null;
  const eventFile = appendQualityReview(a.project, a.plan, a.pipelineId, a.terminalEvent, parsed);
  console.log(JSON.stringify({ op_quality_review: eventFile }));
  runContractGovernor(a.project, a.plan, parsed);
} catch (error) {
  console.error(`Failed to append OpQualityReview: ${error instanceof Error ? error.message : String(error)}`);
  process.exit(1);
}
